### PR TITLE
rlp: refactor interface

### PR DESCRIPTION
**Pull request by ryandotsmith**
encoding example:
```go
item := rlp.List(
	rlp.Byte(4),
	rlp.List(
		rlp.Bytes(src.Ip[:]),
		rlp.Uint16(src.UdpPort),
		rlp.Uint16(src.TcpPort),
	),
	rlp.List(
		rlp.Bytes(dest.Ip[:]),
		rlp.Uint16(dest.UdpPort),
		rlp.Uint16(dest.TcpPort),
	),
	rlp.Time(time.Now().Add(time.Hour)),
)
```
decoding example:
```go
item, err := rlp.Decode(packet[headerSize:])
if err != nil {
        return err
}
hash, err := item.At(1).Hash()
if err != nil {
        return err
}
```

### CODE REVIEW

**Comment by ryandotsmith** [comment]
Here is a link to the entire `Item` implementation: https://github.com/indexsupply/lib/blob/4fa2751a49be184253cb9b8ecfe3435f01d5f225/rlp/rlp.go#L22-L103

An Item is the input to Decode and the output of Encode.

**Review by lechengfan — APPROVED** [review]

**Comment by ryandotsmith** [comment]
Another option is to use reflection and structs to encode/decode. However, I'm less inclined to go this route because I have an aversion to reflection because a) it is too magical for my personal tastes and more importantly b) it has a big --albeit constant-- performance overhead. To further support reason _a_ I would say that the requirement to have struct field definitions be ordered (with respect to the order of an RLP list) is a footgun.

**Comment by ryandotsmith** [comment]
I'm inclined to make `Item.L` and `Item.D` un-exported if this interface seems good.

**Comment by ryandotsmith** [comment]
@kr This is legendary feedback. Amazing. I'm going to digest this for a bit but I wanted to quickly say thank you!

File: rlp/rlp.go [hunk 1]
```diff
@@ -17,6 +19,78 @@ const (
...
+	return &Item{D: buf[4:8]}
+}
+
+func NewUint16(n uint16) *Item {
+	buf := make([]byte, 2)
+	binary.BigEndian.PutUint16(buf[:], n)
+	return &Item{D: buf[:2]}
+}
+
+func (i *Item) Uint16() (uint16, error) {
```
[the remark this answers is not in the document]
Comment (ryandotsmith) [reply rlp/rlp.go:75]: Yes. I should have mentioned that I only threw a couple of types into this PR to demonstrate a proof of concept.

**Comment by ryandotsmith** [comment]
> This is not really about the interface, but more about the internals: did you consider using an interface to represent this as a “sum type” a la http://www.jerf.org/iri/post/2917 ? IMO the current PR is just as reasonable, I’m just curious what your thinking is.

I assume this would apply to the Item's `L` and `D` fields. I was thinking that it would be nice to find a compiler solution to the `L` XOR `D` setup. Ideally you would know at compile that you provided both values and therefore have an error / undetermined state.

File: rlp/rlp.go [hunk 2]
```diff
@@ -17,49 +20,122 @@ const (
...
+func List(items ...Item) Item {
+	if len(items) == 0 {
+		return Item{l: []Item{}}
+	}
+
+	var it Item
+	for i := range items {
+		it.l = append(it.l, items[i])
+	}
+	return it
```
[the remark this answers is not in the document]
Comment (ryandotsmith) [reply rlp/rlp.go]: Oh. I like this, but I do need to deal with an empty list. `List()` should produce `0xc0`.

[the remark this answers is not in the document]
Comment (ryandotsmith) [reply rlp/rlp.go]: I could do something like

```go
if items == nil {
    items = []Item{}
}   
return Item{l: items}
```

File: rlp/rlp.go [hunk 3]
```diff
@@ -17,49 +20,122 @@ const (
...
-
-func Encode(input *Item) ([]byte, error) {
-	if input == nil {
-		return nil, ErrTooFewArgs
-	}
-	if input.D != nil && input.L != nil {
-		return nil, ErrTooManyArgs
+func Encode(input Item) ([]byte, error) {
+	if input.d != nil && input.l != nil {
+		return nil, errors.New("must set d xor l")
```
[the remark this answers is not in the document]
Comment (ryandotsmith) [reply rlp/rlp.go:111]: brilliant! not having an error makes this much simpler!

**Comment by ryandotsmith** [comment]
This was super helpful. I have always wanted to know how to achieve union-like types in Go and now I know how. Thank you.

> Personally to me neither option is clearly superior, so I would choose based on vibes.

I started going down the union-like path but quickly gave up as it seemed to harsh the vibes for this particular case. I think we ended up in a great spot tho. Especially now that Encode doesn't return an error.

File: rlp/rlp.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryandotsmith) [reply rlp/rlp.go]: Yes. I suppose it would be good to define what happens when you call `Bytes(nil)`. Just added a test and fix for that.

**Comment by ryandotsmith** [comment]
rebased to include the length encoding fix

**Pull request merged**